### ISSUE TITLE
self.class.after_commit_counter_cache is nil when calling update from controller

### ISSUE THREAD

**Issue by kosh-jelly** [magnusvk/counter_culture#89]
I'm getting an error when calling update_attributes for an api I'm writing. It's working from the console, but it's not working through the api controller.
```
NoMethodError:
       undefined method `each' for nil:NilClass
```

Here's the full stack trace:

```
counter_culture (0.1.27) lib/counter_culture.rb:249:in `block in _update_counts_after_update&#39;
counter_culture (0.1.27) lib/counter_culture.rb:221:in `call&#39;
counter_culture (0.1.27) lib/counter_culture.rb:221:in `_wrap_in_counter_culture_active&#39;
counter_culture (0.1.27) lib/counter_culture.rb:248:in `_update_counts_after_update&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:424:in `block in make_lambda&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:221:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:221:in `block in halting_and_conditional&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:86:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:86:in `run_callbacks&#39;
activerecord (4.1.2) lib/active_record/callbacks.rb:310:in `_update_record&#39;
activerecord (4.1.2) lib/active_record/timestamp.rb:70:in `_update_record&#39;
activerecord (4.1.2) lib/active_record/persistence.rb:482:in `create_or_update&#39;
activerecord (4.1.2) lib/active_record/callbacks.rb:302:in `block in create_or_update&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:113:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:113:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:166:in `block in halting&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:166:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:166:in `block in halting&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:86:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:86:in `run_callbacks&#39;
activerecord (4.1.2) lib/active_record/callbacks.rb:302:in `create_or_update&#39;
activerecord (4.1.2) lib/active_record/persistence.rb:103:in `save&#39;
activerecord (4.1.2) lib/active_record/validations.rb:51:in `save&#39;
activerecord (4.1.2) lib/active_record/attribute_methods/dirty.rb:21:in `save&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:268:in `block (2 levels) in save&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:329:in `block in with_transaction_returning_status&#39;
activerecord (4.1.2) lib/active_record/connection_adapters/abstract/database_statements.rb:199:in `transaction&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:208:in `transaction&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:326:in `with_transaction_returning_status&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:268:in `block in save&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:283:in `rollback_active_record_state!&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:267:in `save&#39;
activerecord (4.1.2) lib/active_record/persistence.rb:232:in `block in update&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:329:in `block in with_transaction_returning_status&#39;
activerecord (4.1.2) lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `block in transaction&#39;
activerecord (4.1.2) lib/active_record/connection_adapters/abstract/database_statements.rb:209:in `within_new_transaction&#39;
activerecord (4.1.2) lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `transaction&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:208:in `transaction&#39;
activerecord (4.1.2) lib/active_record/transactions.rb:326:in `with_transaction_returning_status&#39;
activerecord (4.1.2) lib/active_record/persistence.rb:230:in `update&#39;
actionpack (4.1.2) lib/action_controller/metal/implicit_render.rb:4:in `send_action&#39;
actionpack (4.1.2) lib/abstract_controller/base.rb:189:in `process_action&#39;
actionpack (4.1.2) lib/action_controller/metal/rendering.rb:10:in `process_action&#39;
actionpack (4.1.2) lib/abstract_controller/callbacks.rb:20:in `block in process_action&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:113:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:113:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:149:in `block in halting_and_conditional&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:166:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:166:in `block in halting&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:229:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:229:in `block in halting&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:166:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:166:in `block in halting&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:86:in `call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:86:in `run_callbacks&#39;
actionpack (4.1.2) lib/abstract_controller/callbacks.rb:19:in `process_action&#39;
actionpack (4.1.2) lib/action_controller/metal/rescue.rb:29:in `process_action&#39;
actionpack (4.1.2) lib/action_controller/metal/instrumentation.rb:31:in `block in process_action&#39;
activesupport (4.1.2) lib/active_support/notifications.rb:159:in `block in instrument&#39;
activesupport (4.1.2) lib/active_support/notifications/instrumenter.rb:20:in `instrument&#39;
activesupport (4.1.2) lib/active_support/notifications.rb:159:in `instrument&#39;
actionpack (4.1.2) lib/action_controller/metal/instrumentation.rb:30:in `process_action&#39;
actionpack (4.1.2) lib/action_controller/metal/params_wrapper.rb:250:in `process_action&#39;
activerecord (4.1.2) lib/active_record/railties/controller_runtime.rb:18:in `process_action&#39;
actionpack (4.1.2) lib/abstract_controller/base.rb:136:in `process&#39;
actionview (4.1.2) lib/action_view/rendering.rb:30:in `process&#39;
rack-mini-profiler (0.9.2) lib/mini_profiler/profiling_methods.rb:108:in `block in profile_method&#39;
actionpack (4.1.2) lib/action_controller/metal.rb:196:in `dispatch&#39;
actionpack (4.1.2) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch&#39;
actionpack (4.1.2) lib/action_controller/metal.rb:232:in `block in action&#39;
actionpack (4.1.2) lib/action_dispatch/routing/route_set.rb:82:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/routing/route_set.rb:82:in `dispatch&#39;
actionpack (4.1.2) lib/action_dispatch/routing/route_set.rb:50:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/journey/router.rb:71:in `block in call&#39;
actionpack (4.1.2) lib/action_dispatch/journey/router.rb:59:in `each&#39;
actionpack (4.1.2) lib/action_dispatch/journey/router.rb:59:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/routing/route_set.rb:678:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/rack/agent_hooks.rb:26:in `traced_call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:55:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/rack/browser_monitoring.rb:23:in `traced_call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:55:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/rack/developer_mode.rb:56:in `traced_call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:55:in `call&#39;
rack (1.5.2) lib/rack/etag.rb:23:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/conditionalget.rb:35:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/head.rb:11:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/params_parser.rb:27:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/flash.rb:254:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/session/abstract/id.rb:225:in `context&#39;
rack (1.5.2) lib/rack/session/abstract/id.rb:220:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/cookies.rb:560:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
activerecord (4.1.2) lib/active_record/query_cache.rb:36:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
activerecord (4.1.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:621:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
activerecord (4.1.2) lib/active_record/migration.rb:380:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call&#39;
activesupport (4.1.2) lib/active_support/callbacks.rb:82:in `run_callbacks&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/callbacks.rb:27:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/reloader.rb:73:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/remote_ip.rb:76:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
railties (4.1.2) lib/rails/rack/logger.rb:38:in `call_app&#39;
railties (4.1.2) lib/rails/rack/logger.rb:20:in `block in call&#39;
activesupport (4.1.2) lib/active_support/tagged_logging.rb:68:in `block in tagged&#39;
activesupport (4.1.2) lib/active_support/tagged_logging.rb:26:in `tagged&#39;
activesupport (4.1.2) lib/active_support/tagged_logging.rb:68:in `tagged&#39;
railties (4.1.2) lib/rails/rack/logger.rb:20:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/request_id.rb:21:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/methodoverride.rb:21:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/runtime.rb:17:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
activesupport (4.1.2) lib/active_support/cache/strategy/local_cache_middleware.rb:26:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/lock.rb:17:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
actionpack (4.1.2) lib/action_dispatch/middleware/static.rb:64:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/sendfile.rb:112:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack-mini-profiler (0.9.2) lib/mini_profiler/profiler.rb:300:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
railties (4.1.2) lib/rails/engine.rb:514:in `call&#39;
railties (4.1.2) lib/rails/application.rb:144:in `call&#39;
newrelic_rpm (3.9.7.266) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call&#39;
rack (1.5.2) lib/rack/lock.rb:17:in `call&#39;
rack (1.5.2) lib/rack/content_length.rb:14:in `call&#39;
rack (1.5.2) lib/rack/handler/webrick.rb:60:in `service&#39;
/Users/josh/.rbenv/versions/2.1.3/lib/ruby/2.1.0/webrick/httpserver.rb:138:in `service&#39;
/Users/josh/.rbenv/versions/2.1.3/lib/ruby/2.1.0/webrick/httpserver.rb:94:in `run&#39;
/Users/josh/.rbenv/versions/2.1.3/lib/ruby/2.1.0/webrick/server.rb:295:in `block in start_thread&#39;
```

Here's what my code looks like:

```
#update method from the controller
 def update
        @domain.update_attributes(domain_params)
end

class ErrorDomain < ActiveRecord::Base
counter_culture [:seed_url, :backlink_seed], 
		column_name: Proc.new {|error_domain| error_domain.available_da? ? 'available_da_count' : nil},
		column_names: {
			['error_domains.available = ? AND error_domains.moz_da > ?', true, 15] => 'available_da_count'
		}

	counter_culture [:seed_url, :backlink_seed], 
		column_name: Proc.new {|error_domain| error_domain.spam_free? ? 'spam_free_count' : nil},
		column_names: {
			["error_domains." + 
				FULL_DATA_COLUMNS.join(" IS NOT NULL AND error_domains.") + " IS NOT NULL" + 
				" AND error_domains.anchor_spam_status IN (?) AND spam_backlinks_count > 0", SPAM_FREE_ANCHOR_STATUSES] => 'spam_free_count'
		}
	counter_culture :seed_url, 
		column_name: Proc.new {|error_domain| error_domain.available_da? ? 'available_da_count' : nil},
		column_names: {
			['error_domains.available = ? AND error_domains.moz_da > ?', true, 15] => 'available_da_count'
		}

	counter_culture :seed_url, 
		column_name: Proc.new {|error_domain| error_domain.spam_free? ? 'spam_free_count' : nil},
		column_names: {
			["error_domains." + 
				FULL_DATA_COLUMNS.join(" IS NOT NULL AND error_domains.") + " IS NOT NULL" + 
				" AND error_domains.anchor_spam_status IN (?) AND spam_backlinks_count > 0", SPAM_FREE_ANCHOR_STATUSES] => 'spam_free_count'
		}
end
```

Apologies for the code dump, but I don't really have an idea what would be causing this. Counter_culture has been working just fine interacting with it from other controllers and background tasks. But it seems to have issues when I'm working with it from the api controller. 

Any chance someone could point me in the right direction? 

Thanks!

**Comment by magnusvk** [comment]
Hmm, very strange. Can you provide a minimal test case that reproduces it? Otherwise it's really hard for me to look at this as I can't reproduce it.

Also, could you please verify this is still happening with the latest gem version?

**Comment by mockdeep** [comment]
I just ran into this issue as well, trying to integrate version `0.1.31`. I implemented it on three models, one of which works fine and the other two break. What the failing ones have in common is that they are single-table inheritance models:

```ruby
class Lesson < ActiveRecord::Base
  has_many :steps
end

class Step < ActiveRecord::Base
  belongs_to :lesson
  counter_culture :lesson
end

class BabyStep < Step
end
```

**Comment by magnusvk** [comment]
Can you fork this and add a failing test to the suite? That would help massively in debugging this. Thanks.

**Comment by mockdeep** [comment]
Sure, this breaks it for me: https://github.com/mockdeep/counter_culture/commit/561519aeb036f59ac3d3a87d05deea577060759b

**Comment by magnusvk** [comment]
What do you think of 71a1417a716bb24aa042640f005b5727a922c082? It fixes your spec, however I'm not quite sure this is the right approach. Would love another set of eyes on it.

**Comment by magnusvk** [comment]
See also #101.

**Comment by magnusvk** [comment]
Now released in 0.1.33
